Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java

```diff
@@ -0,0 +1,467 @@
+                if (!finished) {
+                    throw new RuntimeException("Parallel execution failed");
+                }
+            } catch (InterruptedException e) {
+                throw new RuntimeException("Parallel execution failed", e);
+            }
+        }
+        return new ArrayList<>(Arrays.asList(indexDistancePairListArray));
+    }
+
+    @Override
+    public List<List<Pair<Integer, Double>>> queryAll(int k) {
+        return this.query(this.data, k);
+    }
+
+    /**
+     * Generate a k-d (sub)tree. This method is recursively called to generate the left and right subtrees.
+     * @param d The dimension to be considered for this node of the tree.
+     * @param maxD The maximum dimension of the tree.
+     * @param points The array of points, which are {@link IntAndVector}s of the point's original index and the features.
+     * @param left The lower bound index of for this node of the tree.
+     * @param right The upper bound index of for this node of the tree.
+     * @param distanceType The distance function.
+     * @return The DimensionNode at the root of the (sub)tree being generated.
+     */
+    private static DimensionNode generateTree(int d, int maxD, IntAndVector[] points, int left, int right,
+                                       DistanceType distanceType) {
+        // Handle the termination cases
+        if (right < left) {
+            return null;
+        }
+        if (right == left) {
+            return new DimensionNode(d, points[left], distanceType);
+        }
+
+        // Order the points[left, right] based on the dimension d, such that the mth element will be the median.
+        // Points before it (with lower index) will be <= median, although not sorted. Points after it
+        // (with higher index) will be >= median, again not sorted.
+        int median = 1 + (right-left)/2;
+        setMedian(points, median, left, right, d);
+
+        // Construct a dimension node using the median point
+        DimensionNode medianNode = new DimensionNode(d, points[left+median-1], distanceType);
+
+        // Increment the dimension, resetting back to 0 when it exceeds maxD
+        d++;
+        if (d > maxD) {
+            d = 0;
+        }
+
+        // Generate the below/left and above/right subtrees.
+        medianNode.setBelow(generateTree(d, maxD, points, left, left+median-2, distanceType));
+        medianNode.setAbove(generateTree(d, maxD, points, left+median, right, distanceType));
+        return medianNode;
+    }
+
+    /**
+     * Set the median point for an array of {@link IntAndVector}s based, for a specific dimension, through recursive partitioning
+     * ensuring that points before it (with lower index) will be <= median, although not sorted, and points after it
+     * (with higher index) will be >= median, again not sorted. The order of the array will almost certainly be changed.
```

Comment:
So this says that values equal to the median can be found on both sides of the tree, but that seems to conflict with the documented behaviour of the `partitionOnIndex` function (though not its actual behaviour).